src/components/templates/Poster/usePosterVideo.ts

```diff
@@ -0,0 +1,132 @@
+import { useEffect, useMemo, useState } from "react";
+import Video, { Room, LocalParticipant, RemoteParticipant } from "twilio-video";
+import firebase from "firebase/app";
+
+import { withId, WithId } from "utils/id";
+
+import { User } from "types/User";
+
+import { useUser } from "hooks/useUser";
+import { useWorldUsersById } from "hooks/users";
+
+// TODO: create useVideoRoom hook and divide this into two hooks & refactor it
+export const usePosterVideo = (venueId: string) => {
+  const [room, setRoom] = useState<Room>();
+  const [participants, setParticipants] = useState<
+    Array<LocalParticipant | RemoteParticipant>
+  >([]);
+  const [token, setToken] = useState<string>();
+  const [hasVideo, setHasVideo] = useState(false);
+
+  const { userId } = useUser();
+  const { worldUsersById } = useWorldUsersById();
+
+  const localParticipant = room?.localParticipant;
+
+  useEffect(() => {
+    (async () => {
+      if (!userId) return;
+
+      const getToken = firebase.functions().httpsCallable("video-getToken");
+      const response = await getToken({
+        identity: userId,
+        room: venueId,
+      });
+      setToken(response.data.token);
+    })();
+  }, [venueId, userId]);
+
+  useEffect(() => {
+    if (!localParticipant) return;
+
+    setParticipants((prevParticipants) => [
+      // Hopefully prevents duplicate users in the participant list
+      ...prevParticipants.filter(
+        (p) => p.identity !== localParticipant.identity
+      ),
+      localParticipant,
+    ]);
+  }, [localParticipant]);
+
+  useEffect(() => {
+    if (!token) return;
+
+    const participantConnected = (participant: RemoteParticipant) => {
+      setParticipants((prevParticipants) => [
+        participant,
+        ...prevParticipants.filter((p) => p.identity !== participant.identity),
+      ]);
+    };
+
+    const participantDisconnected = (participant: RemoteParticipant) => {
+      setParticipants((prevParticipants) => {
+        return prevParticipants.filter((p) => p !== participant);
+      });
+    };
+
+    Video.connect(token, {
+      name: venueId,
+      video: hasVideo,
+    }).then((room) => {
```

Comment:
Is how we're using `video` here sufficient? How is it similar/different to what `src/hooks/twilio.ts` does with `setVideoHidden`/etc?

Docs:

- https://media.twiliocdn.com/sdk/js/video/releases/2.13.1/docs/module-twilio-video.html#.connect__anchor
- https://media.twiliocdn.com/sdk/js/video/releases/2.13.1/docs/global.html#ConnectOptions

Reply:
Theoretically, I could reuse that, but the practical problem is that I can only get `Localparticipant | undefined`

![image](https://user-images.githubusercontent.com/63194656/116626513-04ee3880-a954-11eb-98bf-6a2b7a18df36.png)


Which will require the additions to UsePariticipantState, since it only accepts `participant: LocalPariticapnt | RemoteParticipant`

![image](https://user-images.githubusercontent.com/63194656/116626619-38c95e00-a954-11eb-8561-18812733676a.png)

Reply:
> but the practical problem is that I can only get `Localparticipant | undefined`

I don't understand what the practical problem with that is?

> Which will require the additions to UsePariticipantState, since it only accepts participant: LocalPariticapnt | RemoteParticipant

You said you have access to `LocalParticipant` though?

Reply:
I was going to ask/suggest whether we even need to call `Video.connect` here at all when the user isn't going to be on video.. but I guess for them to be able to watch others on video, they are going to need to be yeah?

That's going to mean that we'll be racking up twilio minutes for every user connected in this room, regardless of whether they are currently on video or not. Which could have an outsized impact on our costs/etc depending on the situation.

Is that something we're aware of/have thought through/etc here?

Reply:
> but the practical problem is that I can only get Localparticipant | undefined
> 
> I don't understand what the practical problem with that is?
> 
> Which will require the additions to UsePariticipantState, since it only accepts participant: LocalPariticapnt | RemoteParticipant
> 
> You said you have access to LocalParticipant though?

> 
If it's just that you would need to make the hook also able to accept undefined so that we don't need to try and conditionally render hooks/etc, then that is probably fine to do I think. So long as we ensure that we tell all the appropriate useEffects/etc to essentially be a noop when it's undefined.


Refactored the useParticipant state + moved useVideoRoom into `hooks/twilio` in https://github.com/sparkletown/sparkle/pull/1299/commits/8d7c106d5427dd4489873546fd0809edfe807977

Still think we should leave the implementation as is because 
- With the useParticipant's video turn off, the VideoTrack is still passed + it shows you that the camera is on even though the video is not shown
- I think we should leave the ability to take seat & temporarily turn off the video(during the presentation)

> I was going to ask/suggest whether we even need to call Video.connect here at all when the user isn't going to be on video.. but I guess for them to be able to watch others on video, they are going to need to be yeah?

Yes

> That's going to mean that we'll be racking up twilio minutes for every user connected in this room, regardless of whether they are currently on video or not. Which could have an outsized impact on our costs/etc depending on the situation.
> 
> Is that something we're aware of/have thought through/etc here?

I haven't explicitly mentioned this. Worth mentioning during the next standup though.

> I know we have a max 'currently on video' limit at the moment, but I don't think we have any limit on the number of people that can hang out and watch without being on video?

Frankly, I don't know

Reply:
@mobilevinay @sofisparkle , bringing your awareness ^

Reply:
(Referenced/surfaced the relevant details from above in the main issue in https://github.com/sparkletown/internal-sparkle-issues/issues/546#issuecomment-830989182 as a better place for this to be visible/etc)